pkg/ccl/sqlproxyccl/proxy_handler.go

```diff
@@ -224,15 +226,30 @@ func (handler *proxyHandler) handle(ctx context.Context, proxyConn *conn) error
 
 	var crdbConn net.Conn
 	var outgoingAddress string
-	retryOpts := retry.Options{InitialBackoff: 10 * time.Millisecond, MaxBackoff: time.Second}
+
+	// Repeatedly try to make a connection. Any failures are assumed to be
+	// transient unless the tenant cannot be found (e.g. because it was
+	// deleted). We will simply loop forever, or until the context is canceled
+	// (e.g. by client disconnect). This is preferable to terminating client
+	// connections, because in most cases those connections will simply be
+	// retried, further increasing load on the system.
+	retryOpts := retry.Options{
+		InitialBackoff: 10 * time.Millisecond,
+		MaxBackoff:     5 * time.Second,
+	}
+
 	outgoingAddressErr := log.Every(time.Minute)
 	backendDialErr := log.Every(time.Minute)
 	reportFailureErr := log.Every(time.Minute)
 	var outgoingAddressErrs, codeBackendDownErrs, reportFailureErrs int
+
 	for r := retry.StartWithCtx(ctx, retryOpts); r.Next(); {
+		// Get the DNS/IP address of the backend server to dial.
 		outgoingAddress, err = handler.outgoingAddress(ctx, clusterName, tenID)
```

Comment:
Something has to change here to make `outgoingAddress` work for the case where the directory and the routing rule are both set and the directory doesn't have information about the tenant. In that case the routing rule should be used to construct the address but also we should not loop if unable to connect.

Reply:
I'll be doing this in an upcoming PR.